### PR TITLE
open-scene-graph: various fixes 2015-12

### LINKED ISSUE #43536: Unable to install open-scene-graph: missing dae.h for collada-dom
Trying to install open-scene-graph with:
```brew install --with-collada-dom --with-docs --with-ffmpeg --with-gdal --with-jasper --with-librsvg --with-qt open-scene-graph```

Make fails on 
```[ 95%] Built target osgdb_ive
make: *** [all] Error 2```

My brew logs are here: 
https://gist.github.com/b55823b91e605648ccfc

### LINKED ISSUE #46356: open-scene-graph build fails: iconv link error
I've got an error trying to build open-scene-graph

I'm building with:
```brew install --with-collada-dom --with-dcmtk --with-docs --with-ffmpeg --with-gdal --with-gnuplot --with-jasper --with-librsvg --with-openexr --verbose openscenegraph```

and getting a make error here:
```
[ 72%] Building CXX object src/osgPlugins/shp/CMakeFiles/osgdb_shp.dir/ESRIShape.cpp.o
cd /tmp/open-scene-graph20151125-93815-cie49z/OpenSceneGraph-3.4.0/build/src/osgPlugins/shp && /usr/local/Library/ENV/4.3/clang++   -Dosgdb_shp_EXPORTS -I/tmp/open-scene-graph20151125-93815-cie49z/OpenSceneGraph-3.4.0/include -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks -I/tmp/open-scene-graph20151125-93815-cie49z/OpenSceneGraph-3.4.0/build/include  -std=c++11 -stdlib=libc++  -Wno-conversion -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -fPIC   -o CMakeFiles/osgdb_shp.dir/ESRIShape.cpp.o -c /tmp/open-scene-graph20151125-93815-cie49z/OpenSceneGraph-3.4.0/src/osgPlugins/shp/ESRIShape.cpp
Undefined symbols for architecture x86_64:
  "__libiconv_version", referenced from:
      OFCharacterEncoding::getLibraryVersionString() in libofstd.a(ofchrenc.cc.o)
  "_libiconv", referenced from:
      OFCharacterEncoding::convertString(void*, char const*, unsigned long, OFString&, bool) in libofstd.a(ofchrenc.cc.o)
  "_libiconv_close", referenced from:
      OFCharacterEncoding::closeDescriptor(void*&) in libofstd.a(ofchrenc.cc.o)
  "_libiconv_open", referenced from:
      OFCharacterEncoding::openDescriptor(void*&, OFString const&, OFString const&) in libofstd.a(ofchrenc.cc.o)
  "_libiconvctl", referenced from:
      OFCharacterEncoding::convertString(void*, char const*, unsigned long, OFString&, bool) in libofstd.a(ofchrenc.cc.o)
  "_locale_charset", referenced from:
      OFCharacterEncoding::updateLocaleEncoding() in libofstd.a(ofchrenc.cc.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/osgPlugins-3.4.0/osgdb_dicom.so] Error 1
make[1]: *** [src/osgPlugins/dicom/CMakeFiles/osgdb_dicom.dir/all] Error 2
```

any clues?

### LINKED ISSUE #46372: open-scene-graph build failure on 10.11: c++11-narrowing error
open-scene-graph is failing to build for me on OS X 10.11.1 with Xcode 7.0 or 7.1.

This is for a plain `HOMEBREW_MAKE_JOBS=1 brew install -v --build-from-source open-scene-graph`, with no other options specified.

Build logs: https://gist.github.com/d52bf7d0cd9fa560fff6

```
[ 14%] Building CXX object src/osgDB/CMakeFiles/osgDB.dir/FileUtils.cpp.o
cd /tmp/open-scene-graph20151125-85762-o4wepw/OpenSceneGraph-3.4.0/build/src/osgDB && /usr/local/Library/ENV/4.3/clang++   -DDARWIN_IMAGEIO -DOSGDB_LIBRARY -DOSG_DEFAULT_LIBRARY_PATH=\"/usr/local/Cellar/open-scene-graph/3.4.0/lib/osgPlugins-3.4.0\" -DOSG_PLUGIN_EXTENSION=.so -DUSE_ZLIB -DosgDB_EXPORTS -I/tmp/open-scene-graph20151125-85762-o4wepw/OpenSceneGraph-3.4.0/include -F/Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks -I/tmp/open-scene-graph20151125-85762-o4wepw/OpenSceneGraph-3.4.0/build/include  -std=c++11 -stdlib=libc++  -Wno-conversion -arch x86_64 -isysroot /Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -fPIC   -x objective-c++ -o CMakeFiles/osgDB.dir/FileUtils.cpp.o -c /tmp/open-scene-graph20151125-85762-o4wepw/OpenSceneGraph-3.4.0/src/osgDB/FileUtils.cpp
In file included from /tmp/open-scene-graph20151125-85762-o4wepw/OpenSceneGraph-3.4.0/src/osgDB/FileUtils.cpp:888:
In file included from /Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:80:
In file included from /Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSURLCredential.h:9:
In file included from /Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Security.framework/Headers/Security.h:74:
In file included from /Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Security.framework/Headers/SecStaticCode.h:41:
/Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Security.framework/Headers/CSCommon.h:200:32: error: enumerator value evaluates to -2147483648, which cannot be narrowed to type 'uint32_t' (aka 'unsigned int') [-Wc++11-narrowing]
    kSecCSConsiderExpiration = 1 << 31,         /* consider expired certificates invalid */
                               ^
1 error generated.
make[2]: *** [src/osgDB/CMakeFiles/osgDB.dir/FileUtils.cpp.o] Error 1
make[1]: *** [src/osgDB/CMakeFiles/osgDB.dir/all] Error 2
make: *** [all] Error 2
```

There are existing open issues for open-scene-graph –  #44524 (OS version detection issue), #43536 (collada-dom dependency problem), #46356 (iconv link error) – but they look like they have different causes, so I'm opening a new one.

The weird thing is that this code looks like it's coming from one of Xcode's system-supplied headers, not open-scene-graph. Maybe that header is not clean with respect to the effective warning flags that open-scene-graph is compiling it with. 

`-Wc++11-narrowing` is the warning group that's being triggered here. I don't know what in the formula or brew's superenv definition is enabling it and causing it to be an error.

### PR DESCRIPTION
**Pull request by apjanke**
Multiple fixes to `open-scene-graph`. These fixes are unrelated, but are combined in to a single PR so they can be tested together and get a successful test run. Applying the fixes one by one results in failing tests because there are multiple outstanding issues.

Includes fixes from the following:
Closes #46401
Closes #46388
Closes #46385
Closes #46380

Fixes the following issues:
Fixes #46356
Fixes #43536
Fixes #46372

### CODE REVIEW

File: Library/Formula/open-scene-graph.rb [hunk 1]
```diff
@@ -53,6 +61,10 @@ def install
     args = std_cmake_args
     args << "-DBUILD_DOCUMENTATION=" + ((build.with? "docs") ? "ON" : "OFF")
 
+    # Fixes Xcode 7.1 failure due to warning in system Framework
+    # https://github.com/Homebrew/homebrew/issues/46372
+    args << "-DCMAKE_CXX_FLAGS='-Wno-c++11-narrowing'"
```
Comment (UniqMartin) [inline Library/Formula/open-scene-graph.rb]: Those embedded single quotes are not necessary. Luckily, `cmake` ignores them, so they do not harm. But I'd still vote for getting rid of them.

**Comment by apjanke** [comment]
Test failed. Audit problem, and the `install --devel` is breaking on all OS versions.

I amended to fix the audit failure. Looking in to the `--devel` build failure.

**Comment by apjanke** [comment]
Well, now it's failing to build, only on 10.10.

```
In file included from /tmp/open-scene-graph20151207-23426-rv7y5d/OpenSceneGraph-3.4.0/src/osgDB/FileUtils.cpp:888:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:82:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSURLError.h:14:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/CoreServices.framework/Headers/CoreServices.h:39:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LaunchServices.h:27:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LSInfo.h:529:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LSInfoDeprecated.h:1243:33: error: missing ',' between enumerators
  kLSHandlerOptionsIgnoreCreator __OSX_AVAILABLE_BUT_DEPRECATED_MSG(__MAC_10_4, __MAC_10_11, __IPHONE_4_0, __IPHONE_9_0, "Creator codes are deprecated on OS X.") = 1
                                ^
                                , 
```

**Comment by mikemcquaid** [comment]
Nice work here so far :

**Comment by apjanke** [comment]
Full build failure logs on 10.10: ﻿https://gist.github.com/697a15111e5f5f8fe408

**Comment by apjanke** [comment]
I tried adding `-mmacosx-version-min=10.10` to the CXX flags, since [`Availability.h`](http://www.opensource.apple.com/source/CarbonHeaders/CarbonHeaders-18.1/Availability.h) said the following.

```
For these macros to function properly, a program must specify the OS version range 
    it is targeting.  The min OS version is specified as an option to the compiler:
    -mmacosx-version-min=10.x when building for Mac OS X, and -miphone-version-min=1.x.x
    when building for the iPhone.
```

But it didn't work. Same compilation error.

**Comment by UniqMartin** [comment]
The problem is the 10.11 SDK is using the `__MAC_10_11` macro (and some other related macros) that are not defined in `/usr/include/Availability.h` (the one that ships with OS X 10.10). The same header provided by the 10.11 SDK in `/Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/Availability.h` does define these macros, but as can be seen from the error message, that's not the header that's being used.

This is caused by `superenv` that removes the `-isysroot /Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk` compiler flag. The underlying issues seems to be [that `superenv` believes a sysroot of `/` should be used if the CLT are installed](https://github.com/Homebrew/homebrew/blob/7d714ac258b986a3bde2045c2b2f478973c9c095/Library/Homebrew/extend/ENV/super.rb#L105-L107) while the CMake-based build system of OSG has decided to pick up the 10.11 SDK offered by Xcode, resulting also in the `-F/Applications/Xcode-7.1.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks` flag that is not removed by `superenv`. (It's not entirely clear whether `superenv` or the OSG build system is to blame here.)

(Those `clang++ called with:`, `superenv removed:`, `superenv added:`, and `superenv executed:` lines from [`02.make.cc`](https://gist.github.com/apjanke/697a15111e5f5f8fe408#file-02-make-cc) are quite helpful in understanding what is going on.)

**Comment by apjanke** [comment]
Nice diagnosis. Thanks.

I'll have another look at the OSG CMake and superenv setup and see if I can wrap my head around it better.

**Comment by UniqMartin** [comment]
I think OSG's use of CMake is not necessarily at fault, but more likely `superenv`. #47668 seems to be a closely related (if not identical) issue.

**Comment by apjanke** [comment]
Closing this out as hopeless for the time being.

If I or someone else has time to take another run at this, submit on the new homebrew/core repo.

**Pull request closed without merging**